src/ucs/datastruct/ptr_map.inl

```diff
@@ -0,0 +1,93 @@
+/**
+* Copyright (C) Mellanox Technologies Ltd. 2020.  ALL RIGHTS RESERVED.
+*
+* See file LICENSE for terms.
+*/
+
+#ifndef UCS_PTR_MAP_INL_
+#define UCS_PTR_MAP_INL_
+
+#include "ptr_map.h"
+
+BEGIN_C_DECLS
+
+KHASH_IMPL(ucs_ptr_map_impl, ucs_ptr_map_key_t, void*, 1,
+           kh_int64_hash_func, kh_int64_hash_equal);
+
+static inline ucs_status_t ucs_ptr_map_init(ucs_ptr_map_t *map)
+{
+    map->next_id = 0;
+    kh_init_inplace(ucs_ptr_map_impl, &map->hash);
+    return UCS_OK;
+}
+
+static inline void ucs_ptr_map_destroy(ucs_ptr_map_t *map)
+{
+    kh_destroy_inplace(ucs_ptr_map_impl, &map->hash);
+}
+
+static UCS_F_ALWAYS_INLINE ucs_ptr_map_key_t
+ucs_ptr_map_put(ucs_ptr_map_t *map, void *ptr)
+{
+    ucs_ptr_map_key_t key = (uintptr_t)ptr;
+
+    UCS_STATIC_ASSERT(UCS_PTR_MAP_KEY_INVALID & UCS_PTR_MAP_KEY_ID_FLAG);
+    ucs_assert_always(!(key & UCS_PTR_MAP_KEY_ID_FLAG));
+
+    return key;
+}
+
+static UCS_F_ALWAYS_INLINE ucs_ptr_map_key_t
+ucs_ptr_map_put_safe(ucs_ptr_map_t *map, void *ptr)
```

Comment:
IMO better unite to one function:
- "get" is one function so put should be one function
- for API usage, it's not clear what is the difference between "put_safe" and "put"
 better add a flag like "int indirect" and document its behavior

Reply:
so maybe `unsigned flags` + `UCS_PTR_MAP_PUT_INDIRECT` then?

Reply:
will there be any other flag besides INDIRECT?

Reply:
no need today but who knows for future...

Reply:
maybe add `UCS_PTR_MAP_PUT_DIRECT_PTR` since other possibilities could be container specific e.g. hash, rb_tree, etc

Reply:
i think it would be simpler to just use "int indirect", if smth else is needed can change to flags